espnet/nets/pytorch_backend/frontends/feature_transform.py

```diff
@@ -0,0 +1,235 @@
+from typing import Tuple
+from typing import Union
+
+import librosa
+import numpy as np
+import torch
+from torch_complex.tensor import ComplexTensor
+
+from espnet.nets.pytorch_backend.nets_utils import make_pad_mask
+
+
+class FeatureTransform(torch.nn.Module):
+    def __init__(self,
+                 # Mel options,
+                 fs: int = 16000,
+                 n_fft: int = 512,
+                 n_mels: int = 80,
+                 fmin: float = 0.0,
+                 fmax: float = None,
+
+                 # Normalization
+                 stats_file: str = None,
+                 apply_uttmvn: bool = True,
+                 uttmvn_norm_means: bool = True,
+                 uttmvn_norm_vars: bool = False
+                 ):
+        super().__init__()
+        self.apply_uttmvn = apply_uttmvn
+
+        self.logmel = LogMel(fs=fs, n_fft=n_fft,
+                             n_mels=n_mels, fmin=fmin, fmax=fmax)
+        self.stats_file = stats_file
+        if stats_file is not None:
+            self.global_mvn = GlobalMVN(stats_file)
+        else:
+            self.global_mvn = None
+
+        if self.apply_uttmvn is not None:
+            self.uttmvn = UtteranceMVN(norm_means=uttmvn_norm_means,
+                                       norm_vars=uttmvn_norm_vars)
+        else:
+            self.uttmvn = None
+
+    def forward(self, x: ComplexTensor,
+                ilens: Union[torch.LongTensor, np.ndarray, List[int]]) \
+            -> Tuple[torch.Tensor, torch.LongTensor]:
+        # (B, T, F) or (B, T, C, F)
+        if x.dim() not in (3, 4):
+            raise ValueError(f'Input dim must be 3 or 4: {x.dim()}')
+        if not torch.is_tensor(ilens):
+            ilens = torch.from_numpy(np.asarray(ilens)).to(x.device)
+
+        if x.dim() == 4:
+            # h: (B, T, C, F) -> h: (B, T, F)
+            if self.training:
+                # Select 1ch randomly
+                ch = np.random.randint(x.size(2))
+                h = x[:, :, ch, :]
+            else:
+                # Use the first channel
+                h = x[:, :, 0, :]
+        else:
+            h = x
+
+        # h: ComplexTensor(B, T, F) -> torch.Tensor(B, T, F)
+        h = h.real ** 2 + h.imag ** 2
+
+        h, _ = self.logmel(h, ilens)
+        if self.stats_file is not None:
+            h, _ = self.global_mvn(h, ilens)
+        if self.apply_uttmvn:
+            h, _ = self.uttmvn(h, ilens)
+
+        return h, ilens
+
+
+class LogMel(torch.nn.Module):
+    def __init__(self, fs: int = 16000, n_fft: int = 512, n_mels: int = 80,
+                 fmin: float = None, fmax: float = None):
+        super().__init__()
+
+        _mel_options = dict(sr=fs,
+                            n_fft=n_fft,
+                            n_mels=n_mels,
+                            fmin=fmin,
+                            fmax=fmax)
+        self.mel_options = _mel_options
+
+        # melmat: (D2, D1) -> (D1, D2)
+        melmat = librosa.filters.mel(**_mel_options)
```

Comment:
Is this Mel matrix same as Kaldi's one?
I'm curious.

Reply:
No, it's not.
`htk=True, norm=None` may get the nearest one in librosa, but still different.

https://librosa.github.io/librosa/generated/librosa.core.mel_frequencies.html

Is it a big difference? 
Once I have implemented a fully kaldi-compatible fbank extractor in pure python, https://github.com/nttcslab-sp/kaldifeats, but I think librosa is better because it's well known.

Reply:
This is no problem, but I just want to make sure. I also have some comparisons between them and sometimes Kaldi features are better and sometimes not.
Can you add some comments that this will be different from the one generated from Kaldi?